include/domain_bridge/service_bridge_impl.inc

```diff
@@ -0,0 +1,202 @@
+  const std::string & service_name,
+  typename HackedService<ServiceT>::CallbackT callback,
+  rclcpp::CallbackGroup::SharedPtr group)
+{
+  rcl_service_options_t service_options = rcl_service_get_default_options();
+  auto serv = std::make_shared<HackedService<ServiceT>>(
+    node->get_node_base_interface()->get_shared_rcl_node_handle(),
+    service_name, std::move(callback), service_options);
+  auto serv_base_ptr = std::static_pointer_cast<rclcpp::ServiceBase>(serv);
+  node->get_node_services_interface()->add_service(serv_base_ptr, group);
+  return serv;
+}
+
+rclcpp::Node::SharedPtr
+get_node_for_domain(DomainBridgeImpl & impl, std::size_t domain_id);
+
+const std::string &
+get_node_name(const DomainBridgeImpl & impl);
+}  // namespace detail
+
+/// Bridge a service from one domain to another.
+/**
+ * \param service_bridge: Struct containing info about the service to bridge.
+ * \param options: Options for bridging the topic.
+ */
+template<typename ServiceT>
+void
+DomainBridge::bridge_service(
+  const TopicBridge & service_bridge,
+  const TopicBridgeOptions & options)
+{
+  const auto & node_name = detail::get_node_name(*impl_);
+  // Validate service name
+  const std::string & service_name = rclcpp::expand_topic_or_service_name(
+    service_bridge.topic_name, node_name, "/", true);
+
+  // If a remap name is provided, then validate it
+  // otherwise "remap" to the same name
+  std::string service_remapped = service_name;
+  if (!options.remap_name().empty()) {
+    service_remapped = rclcpp::expand_topic_or_service_name(
+      options.remap_name(), node_name, "/", true);
+  }
+
+  const std::size_t & from_domain_id = service_bridge.from_domain_id;
+  const std::size_t & to_domain_id = service_bridge.to_domain_id;
+
+  // Check if already bridged
+  if (service_bridge_impl_->bridged_services_.find(service_bridge) !=
+    service_bridge_impl_->bridged_services_.end())
+  {
+    std::cerr << "Service '" << service_name << "'" << " already bridged from domain " <<
+      std::to_string(from_domain_id) << " to domain " << std::to_string(to_domain_id) <<
+      ", ignoring" << std::endl;
+    return;
+  }
+
+  rclcpp::Node::SharedPtr from_domain_node = detail::get_node_for_domain(*impl_, from_domain_id);
+  rclcpp::Node::SharedPtr to_domain_node = detail::get_node_for_domain(*impl_, to_domain_id);
+
+  // Create a client for the 'from_domain'
+  auto client = from_domain_node->create_client<ServiceT>(
+    service_name,
+    rmw_qos_profile_services_default,
+    options.callback_group());
+
+  auto handle_request =
+    [client](
+      std::shared_ptr<detail::HackedService<ServiceT>> me,
+      const std::shared_ptr<rmw_request_id_t> & request_header,
+      const std::shared_ptr<typename ServiceT::Request> & request) -> void
+    {
+      // TODO(ivanpauno): What do we do if there's no service server available?
+      // This is a general problem in ROS 2, we don't have a clean way of communicating a service error ....
```

Comment:
IMO, the domain bridge server should mimic the state of the server on the other side of the bridge. E.g. if the target server is not available, then the domain_bridge should not make it's server available. This should help alleviate any issues with client requests queuing up I hope.

Reply:
I understand this may complicate the implementation in the domain_bridge as we'd have to defer the creation of the service server. But it seems like a reasonable expectation from the user. However we ultimately decide, we should add a section about this in the design doc.

Reply:
If we create the service only after detecting the service in the original domain, we will also need to detect when the service goes down and destroy the bridging server.
And TBH that really doesn't solve the issue:

- Server in domain A is up.
- Bridging server in domain B is created.
- Client in domain B makes request.
- Server in domain A goes down.
  - Case 1:
    - We destroy server in domain B, client in domain B never gets response.
      (this is already an issue in the non-bridged case)
  - Case 2:
    - Another server in domain A goes up before we destroy the server in domain B.
    - Client in domain B will never get a reply and never see the server in domain B going down.
      (this is also an issue in the non-bridged case)

I would like to see a way of detecting server failures from the client side in `rclcpp`, e.g.:

```cpp
  client->send_async_request(request, timeout, handle_request_cb, handle_timeout_cb);
```

There's no race there.
If the client wasn't available or didn't repply before the timeout, the timeout callback will be triggered (maybe that one could be called the "error" callback).

(there's other alternatives to handle errors correctly, but the above one is the most straightforward IMO)

---

We can try to make the code more complex here in the meantime .... it will make the races less likely to happen but it won't fix the real issues.

Reply:
I'm don't think we should try and handle the race between a service request and a service going down; this is an issue with services in general, not the domain bridge. 

I'm more concerned with the scenario where the bridge creates a service, even though there is no service to forward requests to. Clients are likely to send requests and never get a response. I think it would be better to only create a service in the bridge once it detects the destination service is available. I would then hope that clients wouldn't have to have special knowledge about how the domain bridge works. E.g. most client code goes something like this:

```c++
client->wait_for_service("foo");
client->send_request(...);
```

If we're dealing with the service bridge as it is currently implemented, then the logic doesn't really make sense since the service being "available" doesn't tell the client anything about the true availability. Does that make sense?

Reply:
Done!

I extended the `WaitForQosHandler` class to also be able to wait for services (it's now names `WaitForGraphEvents`).
I've also refactored the waiting logic a bit in https://github.com/ros2/domain_bridge/pull/26/commits/e5a0eae3a550d51d7bc614a8fa49574556db8f85.

Diff of the commits implementing this: https://github.com/ros2/domain_bridge/pull/26/files/a68b1a811621acd317e868f808c0fba2501881fd..158b8941898c02924312f52e34df2615b1ffe10a.